fuels-core/src/code_gen/abigen.rs

```diff
@@ -180,6 +178,13 @@ impl Abigen {
         let mut seen_struct: Vec<&str> = vec![];
 
         for prop in self.custom_structs.values() {
+            // Skip custom type generation if the custom type is a Sway-native type.
+            // This means ABI methods receiving or returning a Sway-native type
+            // can receive or return that native type directly.
+            if prop.type_field.contains("ContractId") || prop.type_field.contains("Address") {
```

Comment:
Should we also support AssetId? Since a call param is actually using an asset id which is only happens to be a contract id in the case of native tokens.

Reply:
That has me confused a bit. Every Sway case I see in the "wild" is using Sway's `ContractId` (https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/contract_id.sw#L5) to represent an `asset_id` -- which works with the current solution here. I don't even see a proper `AssetId` in the Sway stdlib? But maybe I didn't look deep enough.

But yeah, I can throw in a `prop.type_field.contains("AssetId")` here if `AssetId` is an actual thing in Sway 🤔

Reply:
There is no `AssetId` in Sway, but there is an `AssetId` in Rust! https://github.com/FuelLabs/fuel-types/blob/a852e072c9654b1d50f9495e8ba1de074fb71725/src/types.rs#L254

This should automatically get converted to a `ContractId` as @Voxelot said.

Reply:
Note that in the not-too-distant future Sway will have an `AssetId` as well: https://github.com/FuelLabs/sway/issues/1064

Reply:
what @adlerjohn said.

Reply:
Yeah, I know that we have `AssetId` in Rust. But this line we're commenting on is checking the custom types that were defined in Sway code, coming from the JSON ABI. Actual example:

```
{
  "name": "asset_id",
  "type": "struct ContractId",
  "components": [
    {
      "name": "value",
      "type": "b256",
      "components": null
    }
  ]
},
```

So, here, in this section of the code, we're checking for that `struct ContractId`, which won't be `struct AssetId` until Sway has `AssetId` natively. 

The generated code, then, ends up with `asset_id: ContractId`, e.g.: 

```Rust
#[doc = "Calls the contract\'s `get_balance` (0x000000006fc66a82) function"]
pub fn get_balance(&self,target:ContractId,asset_id:ContractId) -> ContractCall<u64>{
  Contract::method_hash(&self.provider,self.contract_id, &self.wallet,[0,0,0,0,111,198,106,130], &[ParamType::U64], &[target.into_token(),asset_id.into_token(),]).expect("method not found (this should never happen)")
}
```

So I'm guessing that by

> This should automatically get converted to a ContractId as @Voxelot said.

You mean the generated `get_balance`, when receiving `(ContractId, AssetId)` should convert that `AssetId` to `ContractId`? As in:

```Rust
 let id = Contract::deploy(&compiled, &provider, &wallet, TxParameters::default())
    .await
    .unwrap();

// ...

// Turning it into an actually AssetId
let asset = AssetId::from(*id);

// `id` is `ContractId`, asset is `AssetId`
balance_result = instance.get_balance(id, asset).call().await.unwrap();
```

Even though that method was generated as `get_balance(&self,target:ContractId,asset_id:ContractId)`? 

If that's it... It feels a bit hacky to me. Kinda forcing and making up for something that's missing in Sway itself (`AssetId`). 

Though I don't know, am open to anything.

Reply:
I'd be okay with opening an issue for this. If no one complains then it doesn't need to be worked on.